### ISSUE TITLE
Evaluate template & interface preconditions one by one

### ISSUE THREAD

**Issue by cocreature** [digital-asset/daml#11762]
Currently we evaluate all preconditions first and then check the result. I think it makes more sense to switch this to evaluate the preconditions one by one and fail on the first one that doesn’t hold.

**Comment by sofiafaro-da** [comment]
IIRC, template precondition never actually returns false, it throws `PreconditionFailed`. We should make Interface precondition failures do the same thing in LFConversion. That would effectively perform short circuiting, even if not enforced in LF.

**Comment by cocreature** [comment]
> IIRC, template precondition never actually returns false, it throws `PreconditionFailed`. We should make Interface precondition failures do the same thing in LFConversion. That would effectively perform short circuiting, even if not enforced in LF.

that’s true but I still find the current LF evaluation model pretty confusing from a speedy & spec pov so I think I’d prefer the switch anyway. But I don’t feel super strongly about this.

**Comment by sofiafaro-da** [comment]
Thinking more about it, I'm not sure what the right thing to do is. Since exceptions, if a precondition returns false, it's either a bug (... the current state ...), or an attacker. Maybe short-circuit aborting the transaction is the right approach.

We also need to make sure the preconditions are evaluated in a reasonable order, for posterity.

**Comment by cocreature** [comment]
We agreed to do evaluation as follows:
1. Evaluate things one by one and stop as soon as possible.
2. Start with the template precondition
3. Then evaluate the interface conditions in the order they are defined. This needs some changes since we don’t actually preserve that order right now so probably makes sense to do this as a separate step.
